### PR TITLE
feat(benchmark): generate component weights

### LINKED ISSUE #375: dispatch extrinsic weights component calculations

### PR DESCRIPTION
**Pull request by tgntr**
# Description

*Please include a summary of the changes and the related issue. Please also include relevant motivation and context,
including:*

- What does this PR do?
- Why are these changes needed?
- How were these changes implemented and what do they affect?

*Use [Github semantic linking](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) to address any open issues this PR relates to or closes.*

Fixes #375

## Checklist:
- [x] I have read the [contributing guidelines](../docs/docs/contributing/contributing.md).
- [x] My PR title matches the [Conventional Commits spec](https://www.conventionalcommits.org/). 
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] I can start a Substrate node using the updated runtime.

### CODE REVIEW

File: benchmarking/analysis.go
Comment (failfmi) [inline benchmarking/analysis.go:1]: Mismatching types in `analysis_test.go`

File: frame/system/call_remark_weight.go [hunk 1]
```diff
@@ -1,23 +1,17 @@
 // THIS FILE WAS GENERATED USING GOSEMBLE BENCHMARKING PACKAGE
-// DATE: 2024-02-08 12:21:26.168771 +0200 EET m=+34.595230251, STEPS: 50, REPEAT: 20, DBCACHE: 1024, HEAPPAGES: 4096, HOSTNAME: MacBook-Pro.local, CPU: Apple M2 Pro(10 cores, 3504 mhz), GC: extalloc, TINYGO VERSION: 0.31.0-dev, TARGET: polkawasm
+// DATE: 2024-02-27 09:04:21.843885 +0200 EET m=+92.346653293, STEPS: `5`, REPEAT: `2`, DBCACHE: `1024`, HEAPPAGES: 4096, HOSTNAME: MacBook-Pro.local, CPU: Apple M2 Pro(10 cores, 3504 mhz), GC: , TINYGO VERSION: , TARGET:
```
Comment (failfmi) [inline frame/system/call_remark_weight.go]: Can we update all the weights with the new template (specifically for the steps, repeat, dbcache)
Shouldn't heappages be in `` as well?

Comment (failfmi) [reply frame/system/call_remark_weight.go]: I see that the weight generation leaves empty lines, can we trim those?

File: benchmarking/weight_file_extrinsic.go.tmpl [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// THIS FILE WAS GENERATED USING GOSEMBLE BENCHMARKING PACKAGE
+// DATE: {{.Date}}, STEPS: `{{.Steps}}`, REPEAT: `{{.Repeat}}`, DBCACHE: `{{.DbCache}}`, HEAPPAGES: {{.HeapPages}}, HOSTNAME: {{.HostName}}, CPU: {{.CpuName}}, GC: {{.Gc}}, TINYGO VERSION: {{.TinyGoVersion}}, TARGET: {{.Target}}
```
Comment (failfmi) [inline benchmarking/weight_file_extrinsic.go.tmpl]: steps, repeat and dbcache mismatch the overhead template

File: benchmarking/linear_test.go [hunk 3]
```diff
@@ -7,18 +7,22 @@ import (
...
 
-	assert.Equal(t, []uint32{0, 2, 5, 7, 10}, l.values(5))
+	assert.Equal(t, []uint32{0, 2, 5, 7, 10}, l.rangeValues(5))
 
 	l.setValue(99)
 	assert.Equal(t, uint32(99), l.Value())
 
-	componentValues := componentValues([]*linear{l})
-	assert.Equal(t, []uint32{99}, componentValues)
+	// todo
```
Comment (failfmi) [inline benchmarking/linear_test.go]: is this a nit?

File: benchmarking/analysis.go [hunk 4]
```diff
@@ -2,33 +2,62 @@ package benchmarking
...
+// }
+
+// todo may be useless
+//
+//	func (c componentStats) NameCheckUsed() string {
+//		if (c.slope == analysisStats{}) {
+//			return "_"
+//		}
+//		return c.Name()
+//	}
```
Comment (failfmi) [inline benchmarking/analysis.go]: is this still needed?

File: frame/system/call_remark_weight.go (hunk 1 above)
Comment (tgntr) [reply frame/system/call_remark_weight.go]: ty for noticing, i thought empty lines were handled by checking array length>0 within the template. Will fix that

**Review by failfmi — COMMENTED** [review]
Can we update all the previously generated weights with the new format?

**Review by radkomih — DISMISSED** [review]

**Review by failfmi — DISMISSED** [review]

File: polkadot-sdk
Comment (failfmi) [inline polkadot-sdk:1]: nit: wrong submodule hash

**Review by radkomih — DISMISSED** [review]

**Review by failfmi — APPROVED** [review]

**Review by radkomih — APPROVED** [review]

**Pull request merged**